### PR TITLE
Use pickle by default for serialization

### PR DESCRIPTION
**Pull request by pcmoritz**
## Why are these changes needed?

It is still a little while out until we can merge this (i.e. merging the changes that will bring pickling speed up to the current serialization speed on python versions older than 3.8), but this is a first test to see what else needs to be fixed in the CI.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

**Comment by suquark** [comment]
I have run tests locally, here's the list of tests that could have problems:

- python/ray/tests/test_autoscaler_yaml.py
- python/ray/tests/test_component_failures.py
- python/ray/tests/test_failure.py
- python/ray/tests/test_microbenchmarks.py 
- python/ray/tests/test_multi_node.py

**Comment by pcmoritz** [comment]
Thanks Ryans, it does look like most of the failures are due to timeouts etc. I can spend some time working on integrating your pickle5 backport and then we can try that out.

Is there anything you can spot that points to a possible incompatibility (something that is not timeout related)?

File: python/ray/cloudpickle/cloudpickle_fast.py [hunk 1]
```diff
@@ -187,25 +192,46 @@ def _enum_getstate(obj):
...
     """codeobject reducer"""
-    args = (
-        obj.co_argcount, obj.co_posonlyargcount,
-        obj.co_kwonlyargcount, obj.co_nlocals, obj.co_stacksize,
-        obj.co_flags, obj.co_code, obj.co_consts, obj.co_names,
-        obj.co_varnames, obj.co_filename, obj.co_name,
-        obj.co_firstlineno, obj.co_lnotab, obj.co_freevars,
-        obj.co_cellvars
-    )
+    if hasattr(obj, "co_posonlyargcount"):  # pragma: no branch
```
Comment (pcmoritz) [inline python/ray/cloudpickle/cloudpickle_fast.py]: Can you describe why this change was needed?

File: python/ray/cloudpickle/cloudpickle_fast.py [hunk 2]
```diff
@@ -358,14 +383,6 @@ def _function_setstate(obj, state):
     obj.__globals__.update(obj_globals)
     obj.__globals__["__builtins__"] = __builtins__
 
-    if obj_closure is not None:
```
Comment (pcmoritz) [inline python/ray/cloudpickle/cloudpickle_fast.py]: Won't removing this break the pickling if there is a closure?

Comment (suquark) [reply python/ray/cloudpickle/cloudpickle_fast.py]: I have moved this code to where the function will be constructed

File: python/ray/cloudpickle/cloudpickle_fast.py (hunk 1 above)
Comment (suquark) [reply python/ray/cloudpickle/cloudpickle_fast.py]: `co_posonlyargcount` is python3.8+ only

File: python/ray/tests/cluster_utils.py [hunk 3]
```diff
@@ -56,7 +56,8 @@ def connect(self):
         output_info = ray.init(
             ignore_reinit_error=True,
             address=self.redis_address,
-            redis_password=self.redis_password)
+            redis_password=self.redis_password,
+            use_pickle=True)
```
Comment (pcmoritz) [inline python/ray/tests/cluster_utils.py]: This is not actually needed.

**Comment by pcmoritz** [comment]
Jenkins retest this please

**Review by suquark — APPROVED** [review]
LGTM
